pkg/driver/node/mounter/pod_mounter.go

```diff
@@ -0,0 +1,403 @@
+	"k8s.io/klog/v2"
+	"k8s.io/mount-utils"
+
+	"github.com/awslabs/aws-s3-csi-driver/pkg/driver/node/credentialprovider"
+	"github.com/awslabs/aws-s3-csi-driver/pkg/driver/node/envprovider"
+	"github.com/awslabs/aws-s3-csi-driver/pkg/mountpoint"
+	"github.com/awslabs/aws-s3-csi-driver/pkg/podmounter/mountoptions"
+	"github.com/awslabs/aws-s3-csi-driver/pkg/podmounter/mppod"
+	"github.com/awslabs/aws-s3-csi-driver/pkg/util"
+)
+
+const mountpointPodReadinessTimeout = 10 * time.Second
+const mountpointPodReadinessCheckInterval = 100 * time.Millisecond
+
+// mountSyscall is the function that performs `mount` operation for given `target` with given Mountpoint `args`.
+// It returns mounted FUSE file descriptor as a result.
+// This is mainly exposed for testing, in production platform-native function (`mountSyscallDefault`) will be used.
+type mountSyscall func(target string, args mountpoint.Args) (fd int, err error)
+
+// A PodMounter is a [Mounter] that mounts Mountpoint on pre-created Kubernetes Pod running in the same node.
+type PodMounter struct {
+	client                 k8sv1.CoreV1Interface
+	mountpointPodNamespace string
+	mount                  mount.Interface
+	kubeletPath            string
+	mountSyscall           mountSyscall
+	kubernetesVersion      string
+	credProvider           *credentialprovider.Provider
+}
+
+// NewPodMounter creates a new [PodMounter] with given Kubernetes client.
+func NewPodMounter(client k8sv1.CoreV1Interface, credProvider *credentialprovider.Provider, mountpointPodNamespace string, mount mount.Interface, mountSyscall mountSyscall, kubernetesVersion string) (*PodMounter, error) {
+	return &PodMounter{
+		client:                 client,
+		mountpointPodNamespace: mountpointPodNamespace,
+		mount:                  mount,
+		kubeletPath:            util.KubeletPath(),
+		mountSyscall:           mountSyscall,
+		kubernetesVersion:      kubernetesVersion,
+		credProvider:           credProvider,
+	}, nil
+}
+
+// Mount mounts the given `bucketName` at the `target` path using provided credential context and Mountpoint arguments.
+//
+// At high level, this method will:
+//  1. Wait for Mountpoint Pod to be `Running`
+//  2. Write credentials to Mountpoint Pod's credentials directory
+//  3. Obtain a FUSE file descriptor
+//  4. Call `mount` syscall with `target` and obtained FUSE file descriptor
+//  5. Send mount options (including FUSE file descriptor) to Mountpoint Pod
+//  6. Wait until Mountpoint successfully mounts at `target`
+//
+// If Mountpoint is already mounted at `target`, it will return early at step 2 to ensure credentials are up-to-date.
+func (pm *PodMounter) Mount(ctx context.Context, bucketName string, target string, credentialCtx credentialprovider.ProvideContext, args mountpoint.Args) error {
+	podID, volumeID := credentialCtx.PodID, credentialCtx.VolumeID
+
+	err := pm.checkTargetPath(target)
+	if err != nil {
+		return fmt.Errorf("Failed to check target path %q: %w", target, err)
```

Comment:
"Failed to verify target path can be used as a mount point"?

Reply:
Addressed with https://github.com/awslabs/mountpoint-s3-csi-driver/pull/368/commits/b910e556b706e0a207c4c7e0e194e0d02c3b7085